File: app/src/main/java/com/isl/api/ApiClient.java
package com.isl.api;

import android.content.Context;
import android.os.Build;
import android.util.Log;

import androidx.annotation.RequiresApi;

import com.isl.dao.DataBaseHelper;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ApiClient {
    private static IApiRequest iApiRequest;
    private static String url;


    @RequiresApi(api = Build.VERSION_CODES.N)
    public static void init(Class<IApiRequest> requestClass, Context context) {
        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        DataBaseHelper db = new DataBaseHelper(context);
        db.open();
        url = db.getModuleIP("Audits");
        if (url.contains("~")) {
            url = url.substring(0, url.indexOf("~"));
        }
        if (url.contains("http") || url.contains("https")) {

        } else {
            //url="https://"+url+"/api/";
            url = "https://" + url + "/";
        }
        Log.d("URL_IP-->", url);


        db.close();

        OkHttpClient.Builder okclient = new OkHttpClient.Builder();
        okclient.connectTimeout(120, TimeUnit.SECONDS);
        okclient.readTimeout(120, TimeUnit.SECONDS);
        okclient.writeTimeout(120, TimeUnit.SECONDS);

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(url)
                .addConverterFactory(GsonConverterFactory.create())
                .client(okclient.build())
                .build();

        iApiRequest = retrofit.create(requestClass);
    }

    public static IApiRequest getRequest() {
        return iApiRequest;
    }
}

File: app/src/main/java/com/isl/leaseManagement/api/ApiClient.kt
package com.isl.leaseManagement.api

import ApiInterceptor
import IApiRequest
import com.google.gson.GsonBuilder
import com.isl.leaseManagement.base.BaseActivity
import com.isl.leaseManagement.utils.AppConstants
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import org.json.JSONObject
import retrofit2.HttpException
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

object ApiClient {  //this class is  for creating an instance which will call apis

    var request: IApiRequest? = null

    fun init() {
        val interceptor = HttpLoggingInterceptor()
        interceptor.level = HttpLoggingInterceptor.Level.BODY
        val okHttpClient = OkHttpClient.Builder()
            .addNetworkInterceptor(ApiInterceptor())
            .addInterceptor(interceptor)
            .connectTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .build()


        val retrofit = Retrofit.Builder()
            .baseUrl(AppConstants.baseUrl)
            .addConverterFactory(
                GsonConverterFactory.create(
                    GsonBuilder().serializeNulls().create()
                )
            )
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .client(okHttpClient)
            .build()

        request = retrofit.create(IApiRequest::class.java)
    }

    fun handleApiError(e: Throwable, defaultErrorMessage: String, baseActivity: BaseActivity) {
        if (e is HttpException) {
            try {
                val errorBody = e.response()?.errorBody()?.string()
                if (!errorBody.isNullOrEmpty()) {
                    val jsonError = JSONObject(errorBody)

                    // First, try to parse the single error message format
                    if (jsonError.has("error") && jsonError.getJSONObject("error").has("message")) {
                        val errorMessage = jsonError.getJSONObject("error").getString("message")
                        baseActivity.showToastMessage(errorMessage)
                    }
                    // If not, try to parse the array of errors
                    else if (jsonError.has("errors")) {
                        val errorsArray = jsonError.getJSONArray("errors")
                        if (errorsArray.length() > 0) {
                            val firstError = errorsArray.getJSONObject(0).getString("message")
                            baseActivity.showToastMessage(firstError)
                        } else {
                            baseActivity.showToastMessage(defaultErrorMessage)
                        }
                    }
                    // Fallback to the default message if neither format is found
                    else {
                        baseActivity.showToastMessage(defaultErrorMessage)
                    }
                } else {
                    baseActivity.showToastMessage(defaultErrorMessage)
                }
            } catch (jsonException: Exception) {
                baseActivity.showToastMessage(defaultErrorMessage)
            }
        } else {
            baseActivity.showToastMessage(defaultErrorMessage)
        }
    }



}